### PR TITLE
kernel: refactor implications code

### PR DESCRIPTION
**Pull request by fingolfin**
Pull out some duplicate code implementing a simple hashtable; this doesn't really save much code, but it makes the functions using the new helpers (`FuncWITH_HIDDEN_IMPS_FLAGS` and `FuncWITH_IMPS_FLAGS`) more readable.

### CODE REVIEW

**Review by jackschmidt — COMMENTED** [review]
Thanks! The new version is a lot easier to read. There is at least one more "inline" hash table implementation just in this file, and comparing the two, the split out version is much easier to follow and reason about.

For my review, I mostly just point out the tricky parts, which as far as I can tell were done correctly, but perhaps it will make it easy for experienced eyes.

I don't think I have permission to edit the file with the suggested change (the variable scope), but that is mostly just style anyways.

File: src/opers.cc [hunk 1]
```diff
@@ -638,6 +638,46 @@ static Obj FuncAND_FLAGS(Obj self, Obj flags1, Obj flags2)
...
+
+        if (!old_flags)
+            break;
+
+        new_flags = old_flags;
+        new_with = old_with;
+        hash = (hash * 311 + 61) % len;
+    }
+
+    CHANGED_BAG(ht);
```
[the remark this answers is not in the document]
Comment (fingolfin) [reply src/opers.cc:677]: Yes, we know that; `ELM_PLIST` simply reads a word from memory. (Note that this code was there before and just moved around)

File: src/opers.cc [hunk 2]
```diff
@@ -686,7 +726,6 @@ static Obj FuncCLEAR_HIDDEN_IMP_CACHE(Obj self, Obj filter)
     {
         SET_ELM_PLIST(WITH_HIDDEN_IMPS_FLAGS_CACHE, i, 0);
         SET_ELM_PLIST(WITH_HIDDEN_IMPS_FLAGS_CACHE, i + 1, 0);
-        CHANGED_BAG(WITH_HIDDEN_IMPS_FLAGS_CACHE);
```
[the remark this answers is not in the document]
Comment (fingolfin) [reply src/opers.cc:689]: Exactly.

File: src/opers.cc [hunk 3]
```diff
@@ -713,29 +752,23 @@ static Obj FuncWITH_HIDDEN_IMPS_FLAGS(Obj self, Obj flags)
 #endif
     Int changed, i, lastand, stop;
     Int hidden_imps_length = LEN_PLIST(HIDDEN_IMPS) / 2;
-    Int base_hash = INT_INTOBJ(FuncHASH_FLAGS(0, flags)) % HIDDEN_IMPS_CACHE_LENGTH;
-    Int hash = base_hash;
-    Int hash_loop = 0;
+    Int hash =
+        INT_INTOBJ(FuncHASH_FLAGS(0, flags)) % HIDDEN_IMPS_CACHE_LENGTH;
     Obj cacheval;
```
[the remark this answers is not in the document]
Comment (fingolfin) [reply src/opers.cc:757]: In general it's considered bad style in modern C to write `if (var = func())` and compilers will warn about it; one has to write `if ((var = func()))` to silence the warnings.

Oh and `if (int var = func())`  is only valid in C++, not C (of course this file is a C++ file but in general we try to keep the C++-isms to a minimum)

**Review by wilfwilson — APPROVED** [review]
Looks alright to me.

**Comment by ChrisJefferson** [comment]
This seems to have slowed GAP's startup time by over 3x for me (from 1.5s to 5s)

**Comment by fingolfin** [comment]
Ouch!! I'll investigate (or maybe open an issue so i don't forget, am at lunch now)

File: src/opers.cc [hunk 4]
```diff
@@ -757,28 +790,9 @@ static Obj FuncWITH_HIDDEN_IMPS_FLAGS(Obj self, Obj flags)
...
-    hash = base_hash;
-
-    new_with = with;
-    new_flags = flags;
-
-    for (hash_loop = 0; hash_loop < 3; ++hash_loop) {
-        old_flags = ELM_PLIST(WITH_HIDDEN_IMPS_FLAGS_CACHE, hash * 2 + 1);
-        old_with = ELM_PLIST(WITH_HIDDEN_IMPS_FLAGS_CACHE, hash * 2 + 2);
+    StoreHashTable<HIDDEN_IMPS_CACHE_LENGTH>(WITH_HIDDEN_IMPS_FLAGS_CACHE,
+                                             hash, with, flags);
```
Comment (ChrisJefferson) [inline src/opers.cc:794]: you pass in 'hash, with, flags' here (and other places), but the new function takes 'hash, flags, with).

Comment (fingolfin) [reply src/opers.cc:794]: Oh wow. It's horrifying that this even runs

**Pull request merged**